python/ray/ml/utils/remote_storage.py

```diff
@@ -0,0 +1,142 @@
+import re
+from typing import Optional, Tuple
+
+try:
+    import fsspec
+except ImportError:
+    fsspec = None
+
+try:
+    import pyarrow
+    import pyarrow.fs
+except (ImportError, ModuleNotFoundError):
+    pyarrow = None
+
+from ray import logger
+
+URI_PROTOCOL_REGEX = r"^(\w+)://(.+)$"
+S3_PREFIX = "s3://"
+GS_PREFIX = "gs://"
+HDFS_PREFIX = "hdfs://"
+ALLOWED_REMOTE_PREFIXES = (S3_PREFIX, GS_PREFIX, HDFS_PREFIX)
+
+
+def _assert_pyarrow_installed():
+    if pyarrow is None:
+        raise RuntimeError(
+            "Uploading, downloading, and deleting from cloud storage "
+            "requires pyarrow to be installed. Install with: "
+            "`pip install pyarrow`. Subsequent calls to cloud operations "
+            "will be ignored."
+        )
+
+
+def _split_protocol_path(uri: str) -> Tuple[Optional[str], Optional[str]]:
+    protocol_match = re.match(URI_PROTOCOL_REGEX, uri)
+    if not protocol_match:
+        return None, None
+
+    return protocol_match.group(1), protocol_match.group(2)
+
+
+def fs_hint(uri: str) -> str:
+    """Return a hint how to install required filesystem package"""
+    if pyarrow is None:
+        return "Please make sure PyArrow is installed: `pip install pyarrow`."
+    if fsspec is None:
+        return "Try installing fsspec: `pip install fsspec`."
+
+    from fsspec.registry import known_implementations
+
+    protocol, _path = _split_protocol_path(uri)
+
+    if protocol in known_implementations:
+        return known_implementations[protocol]["err"]
+
+    return "Make sure to install and register your fsspec-compatible filesystem."
+
+
+def is_cloud_target(uri: str) -> bool:
+    """Check if target URI is a cloud target"""
+    if bool(get_fs_and_path(uri)[0]):
+        return True
+    # Keep manual check for prefixes for backwards compatibility with the
+    # TrialCheckpoint class. Remove once fully deprecated.
+    if any(uri.startswith(p) for p in ALLOWED_REMOTE_PREFIXES):
+        return True
+    return False
+
+
+def get_fs_and_path(
+    uri: str,
+) -> Tuple[Optional["pyarrow.fs.FileSystem"], Optional[str]]:
+    if not pyarrow:
+        return None, None
+
+    try:
+        fs, path = pyarrow.fs.FileSystem.from_uri(uri)
```

Comment:
Just curious, what kind of uri will go through line 77 v.s. line 97?

Reply:
line 77: built-in pyarrow filesystems
line 97: wrapped fsspec filesystems

Reply:
got it, thanks!